plugins/module_utils/crypto/cryptography_openssh.py

```diff
@@ -0,0 +1,484 @@
+
+    _TEXT_ENCODING = 'UTF-8'
+
+    class InvalidAlgorithmError(Exception):
+        pass
+
+    class InvalidCommentError(Exception):
+        pass
+
+    class InvalidDataError(Exception):
+        pass
+
+    class InvalidKeyFileError(Exception):
+        pass
+
+    class InvalidKeySizeError(Exception):
+        pass
+
+    class InvalidKeyTypeError(Exception):
+        pass
+
+    class InvalidPassphraseError(Exception):
+        pass
+
+    class Asymmetric_Keypair(object):
+        """Container for newly generated asymmetric keypairs or those loaded from existing files
+
+           :path: A path to an existing private key to be loaded
+           :keytype: One of rsa, dsa, ecdsa, ed25519
+           :size: The key length for newly generated keys
+           :passphrase: Secret used to encrypt or decrypt a new or existing key respectively
+        """
+
+        def __init__(self, path=None, keytype='rsa', size=None, passphrase=None):
+
+            self.__algorithm_parameters = {
+                'rsa': {
+                    'default_size': 2048,
+                    'valid_sizes': range(1024, 16384),
+                },
+                'dsa': {
+                    'default_size': 1024,
+                    'valid_sizes': [1024],
+                },
+                'ed25519': {
+                    'default_size': 256,
+                    'valid_sizes': [256],
+                },
+                'ecdsa': {
+                    'default_size': 256,
+                    'valid_sizes': [256, 384, 521],
+                    'curves': {
+                        256: ec.SECP256R1(),
+                        384: ec.SECP384R1(),
+                        521: ec.SECP521R1(),
+                    }
+                }
+            }
+
+            self.__privatekey = None
+            self.__publickey = None
+
+            if passphrase:
+                passphrase_validation_error = self.__validate_passphrase(passphrase)
+
+                if passphrase_validation_error:
+                    raise InvalidPassphraseError("%s cannot be encoded from text" % passphrase)
+
+                self.__encryption_algorithm = serialization.BestAvailableEncryption(passphrase.encode(encoding=_TEXT_ENCODING))
+            else:
+                self.__encryption_algorithm = serialization.NoEncryption()
+
+            if path:
+                self.__keytype = None
+                self.__load_keypair(path, passphrase)
+            else:
+                if keytype not in self.__algorithm_parameters.keys():
+                    raise InvalidKeyTypeError("%s is not a valid keytype. Valid keytypes are %s" % (keytype, ", ".join(self.__algorithm_parameters.keys())))
+
+                self.__keytype = keytype
+
+                if size:
+                    self.__size = size
+                else:
+                    self.__size = self.__algorithm_parameters[self.__keytype]['default_size']
+
+                self.__generate_keypair()
```

Comment:
I think it would be better to have class methods with speaking names to do these tasks, i.e. `Asymmetric_Keypair.load()` to load a keypair and `Asymmetric_Keypair.generate()` to generate one.

Reply:
Have you thought about this one?

Reply:
I did remove _keypair from these method names, but I may be missing the meaning of your suggestion.
Was the intention that these be readable as _actions_ solely in the context of initialization or that the methods also be exposed for end users? I intentionally wanted to present asymmetric keypairs as immutable single-use objects hence why the loading/generating mechanisms are obscured.

Reply:
I would offer them as class methods, i.e. you can do `AsymmetricKeypair.load(filename)` to load a keypair, which will return an `AsymmetricKeypair` instance. This is more expressive then using `AsymmetricKeypair(path=filename)` IMO.

Reply:
(See for example https://code-maven.com/slides/python/class-methods-alternative-constructor for this kind of pattern.)

Reply:
Ok, that does open the possibility of an _empty_ object and makes the faux overloaded constructor less confusing. I will spend some time on that. Thanks

Reply:
You could declare the constructor private (by docstring, no idea whether Python allows to do that explicitly without using ugly hacks) to avoid having an empty keypair.

Reply:
I am considering the possibility of an empty object a benefit.
That would allow for the truthiness of the class to be evaluated in a more obvious (pythonic) way.

Reply:
Hmm, I think it's a lot easier to let only `None` represent an empty key, instead of having an instance of the class which can represent an empty key. While such an empty value could simplify some cases, it in general makes code more complicated, since it is harder to ensure that a key is actually specified. (For example, it is basically impossible to specify this requirement with the type system, which is IMO a huge disadvantage.)